astropy/table/table.py

```diff
@@ -785,35 +785,55 @@ def _new_from_slice(self, slice_):
         table.meta.clear()
         table.meta.update(deepcopy(self.meta))
         table.primary_key = self.primary_key
-        cols = self.columns.values()
 
         newcols = []
-        for col in cols:
-            col.info._copy_indices = self._copy_indices
+        for col in self.columns.values():
             newcol = col[slice_]
-            if col.info.indices:
+
+            # Note in line below, use direct attribute access to col.indices for Column
+            # instances instead of the generic col.info.indices.  This saves about 4 usec
+            # per column.
+            if (col if isinstance(col, Column) else col.info).indices:
+                # TODO : as far as I can tell the only purpose of setting _copy_indices
+                # here is to communicate that to the initial test in `slice_indices`.
+                # Why isn't that just sent as an arg to the function?
+                col.info._copy_indices = self._copy_indices
                 newcol = col.info.slice_indices(newcol, slice_, len(col))
+
+                # Don't understand why this is forcing a value on the original column.
+                # Normally col.info does not even have a _copy_indices attribute.  Tests
+                # still pass if this line is deleted.  (Each col.info attribute access
+                # is expensive).
+                col.info._copy_indices = True
+
             newcols.append(newcol)
-            col.info._copy_indices = True
 
-        self._make_table_from_cols(table, newcols)
+        self._make_table_from_cols(table, newcols, check_cols=False,
+                                   names=self.columns.keys())
         return table
 
     @staticmethod
-    def _make_table_from_cols(table, cols):
+    def _make_table_from_cols(table, cols, check_cols=True, names=None):
         """
         Make ``table`` in-place so that it represents the given list of ``cols``.
         """
-        colnames = set(col.info.name for col in cols)
-        if None in colnames:
-            raise TypeError('Cannot have None for column name')
-        if len(colnames) != len(cols):
-            raise ValueError('Duplicate column names')
+        if check_cols:
+            colnames = set(col.info.name for col in cols)
+            if None in colnames:
+                raise TypeError('Cannot have None for column name')
+            if len(colnames) != len(cols):
+                raise ValueError('Duplicate column names')
+
+        if names is None:
```

Comment:
I would move this above `if check_cols`, so that you can use it in the `check_cols` clause, i.e.,
```
if names is None:
   ...
elif len(names) != len(cols):
   raise ...  # omit this?

if check_cols:
    if None in names:
        ...
   if len(set(names)) != len(cols):
        ...
```
If going for that, you could also remove `check_cols` to the end (or just omit it?? the main delay surely is in getting the names!). 

Finally, how about calling it `verify`?